File: src/graphs2/dag_sssp.cpp
// DAG 最短路徑（CLRS §22.2）：拓撲排序後按序單趟鬆弛——負權 OK、Θ(V+E)
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

const long long INF = 1e18;
std::vector<std::vector<std::pair<int,int>>> adj;   // {v, w}
std::vector<int> order;
std::vector<bool> visited;

void visit(int u) {
    visited[u] = true;
    for (auto [v, w] : adj[u])
        if (!visited[v]) visit(v);
    order.push_back(u);                     // 完成時 push → 反轉即拓撲序
}

int main() {
    const std::string name = "rstxyz";      // CLRS 圖 22.5
    adj = {{{1,5},{2,3}}, {{2,2},{3,6}}, {{3,7},{4,4},{5,2}},
           {{4,-1},{5,1}}, {{5,-2}}, {}};
    int n = adj.size();
    visited.assign(n, false);
    for (int u = 0; u < n; ++u)
        if (!visited[u]) visit(u);
    std::reverse(order.begin(), order.end());

    std::vector<long long> d(n, INF);
    d[1] = 0;                               // 源點 s
    for (int u : order) {                   // 按拓撲序，每邊恰鬆弛一次
        if (d[u] == INF) continue;
        for (auto [v, w] : adj[u])
            if (d[u] + w < d[v]) d[v] = d[u] + w;
    }
    std::cout << "DAG-SHORTEST-PATHS（源點 s）：";
    for (int v = 0; v < n; ++v)
        std::cout << name[v] << ":" << (d[v] == INF ? std::string("∞") : std::to_string(d[v])) << " ";
    std::cout << "\n";
    return 0;
}

File: src/graphs2/dag_sssp.py
# DAG 最短路徑（CLRS §22.2，改寫自官方 dag_shortest_paths.py，MIT License）
# 拓撲排序後「按序單趟鬆弛」——允許負權、Θ(V+E)、比 Dijkstra 還快
INF = float("inf")
adj = {"r": [("s",5),("t",3)], "s": [("t",2),("x",6)],
       "t": [("x",7),("y",4),("z",2)],
       "x": [("y",-1),("z",1)], "y": [("z",-2)], "z": []}   # CLRS 圖 22.5


def topological_sort(adj):
    fin, visited = [], set()

    def visit(u):
        visited.add(u)
        for v, _ in adj[u]:
            if v not in visited:
                visit(v)
        fin.append(u)

    for u in adj:
        if u not in visited:
            visit(u)
    return list(reversed(fin))              # 完成時間反序


def dag_shortest_paths(adj, s):
    d = {v: INF for v in adj}
    d[s] = 0
    for u in topological_sort(adj):          # 按拓撲序
        if d[u] == INF:
            continue
        for v, w in adj[u]:                  # 每條邊恰鬆弛一次
            if d[u] + w < d[v]:
                d[v] = d[u] + w
    return d


d = dag_shortest_paths(adj, "s")
print("DAG-SHORTEST-PATHS（源點 s）：",
      " ".join("%s:%s" % (v, "∞" if d[v] == INF else d[v]) for v in "rstxyz"))
